django/test/utils.py

```diff
@@ -401,18 +402,31 @@ def enable(self):
         self.wrapped = settings._wrapped
         settings._wrapped = override
         for key, new_value in self.options.items():
-            setting_changed.send(sender=settings._wrapped.__class__,
-                                 setting=key, value=new_value, enter=True)
+            try:
+                setting_changed.send(sender=settings._wrapped.__class__,
+                                     setting=key, value=new_value, enter=True)
+            except Exception as ex:
+                self.enable_exception = ex
+                self.disable()
```

Comment:
This will cause `disable` to be called twice if `override_settings` is used as a context decorator because `TestDecorator.__exit__` always executes `disable()`

https://github.com/slafs/django/blob/0d9284b305f417f584a9ed4fd1ca8397a1eed49d/django/test/utils.py#L338-L339

This is an issue because this will cause the `setting_changed(enter=False)` signals to be fired twice and the traceback will be really odd to users as it will be of the form

`__enter__` -> `enable` -> `disable` -> `__exit__` -> `disable` -> `(signal receiver trace)`

It should be off the form `__enter__` -> `enable` -> `(signal receiver trace)`

You should be able to assert this works correctly by inspecting the `__cause__` and `__context__` attributes of exceptions.

Reply:
One solution here would be to make `enable`/`disable` idempotent or to override `__exit__` to deal with `(exc_type, exc_value, traceback) == self.enable_exception` in a special way.

Reply:
I don't think I agree with you here or I'm missing something.

> This will cause disable to be called twice if override_settings is used as a context decorator because TestDecorator.__exit__ always executes disable()
https://github.com/slafs/django/blob/0d9284b305f417f584a9ed4fd1ca8397a1eed49d/django/test/utils.py#L338-L339

I'm pretty sure, that's not true. Calling `disable` here will eventually raise an exception (stored inside `enable_exception` var) and when `__enter__` doesn't return cleanly (like in this case), `__exit__` is not called at all.

I guess the rest of your comment(s) was based on that initial assumption, which I'm disagreeing with, so I'm not going to refer to it for now.

Reply:
> I'd also be great to have tests for nested settings overrides and exceptions raised during a context manager and disabling.

Please let me know if you still want me to add tests for that, in the context of my comment above.

Reply:
You are right I missed the fast that exceptions raised during `__enter__` are not going to go through `__exit__` sorry for that. That makes these tests unnecessary.